### DISCUSSION TITLE
Release 2.1.7 Improvements to Simple Volumetric Model (SVM)

### DISCUSSION THREAD

**Discussion by gwlucastrig** [gwlucastrig/Tinfour#78] [General]
The majority of the work for release 2.1.7 was directed to improving the Simple Volumetric Model (SVM).  This model uses the Delaunay triangulation to model reservoir capacity with respect to changing water levels.  Improvements to the module included better integration with standard GIS utilities, including the ability to export GIS-compatible contours in shapefile format and raster data in floating-point grid format. In support of these changes, the GIS module was improved for better shapefile support.

A number of other bugs and code clean up changes were also implemented.

<hr /><em>This discussion was created from the release <a href='https://github.com/gwlucastrig/Tinfour/releases/tag/'>Release 2.1.7 Improvements to Simple Volumetric Model (SVM)</a>.</em>

**Comment by sonw7** [comment]
After importing models and constraints, how can the results be exported as shape files

**Reply by gwlucastrig** [reply]
I've made some comments in the discussion below.

**Comment by gwlucastrig** [comment]
I'm not sure how far along you are working with the SVM, but if you've created an SVM properties file for your project, then you can use it to specify shapefile outputs.

The main discussion of the SVM setup is at https://gwlucastrig.github.io/TinfourDocs/HydrographySimpleVolumetricModel/index.html
But, that discussion kind of omits a discussion of the output shapefiles.  So what you could do is look through the sample properties file and find the lines: 
 
    # The shapefile specification indicate the names of shapefiles used to write
    # polygon (_r for "region") and/or line-based features (_l for "line").
    # The .prj file from the source data is transcribed to make prj files
    # for the output.

    #contourRegionShapefile = contour/SvmContourAlanHenry_r.shp
    #contourLineShapefile   = contour/SvmContourAlanHenry_l.shp
 
In the example file, I was dealing with Lake Alan Henry, so I named the shapefiles accordingly.  You will probably want to pick names more suitable to your own project.  And, of course, you'll want to uncomment the two specifications...

I hope this gets you started.  Feel free to post more questions if you encounter any more roadblocks.  

Also, I'd love to hear more about what you are doing with the SVM.  I haven't received any feedback from users, so I haven't had a lot of information in terms of improving the model.
 
Gary

**Reply by sonw7** [reply]
Thank you for your reply, which is very helpful to me! There are still some questions I want to ask you. After I run tinforviewmain, I import the model from the shape file to form Delaunay triangulation, and then import constraints from the shape file. After importing constraints, I can't find the relevant algorithm of this process, and the data structure related to triangles in the storage grid can't be found. Can you direct me where I can find it?

**Reply by gwlucastrig** [reply]
I'm not sure I understand what you're looking for.  But I'll try to supply useful information. The TInfour Viewer is a demonstration program. It is intended to give software developers ideas of how to use the Tinfour API, but it's not an especially useful program.  The main class for the viewer is TinfourViewerMain.java.

You may find it easier to look at the ExampleGridAndHillshade.java class which accepts some inputs and constraints and uses them to produce a hillshade image.  The example class is intended to show how to use the API.  There is also some discussion related to it in the document https://gwlucastrig.github.io/TinfourDocs/TinfourAlgorithmsAndDataElements.pdf starting at around page 44 which discusses  the test environment.   When you chose an interpolation class, I'd recommend the Natural Neighbor Interpolation which is really the single best implementation in the Tinfour library.

Hope this helps.

Gary
